jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/util/NonBindingJettyMethodHandle.java

```diff
@@ -0,0 +1,133 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal.util;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.WrongMethodTypeException;
+import java.util.Objects;
+
+class NonBindingJettyMethodHandle implements JettyMethodHandle
+{
+    private final MethodHandle _methodHandle;
+    private final Object[] _parameters;
+    private final Boolean[] _boundParams;
+    private final int _numParams;
+    private Class<Object> _returnType;
+    private MethodHandle _returnFilter;
+
+    public NonBindingJettyMethodHandle(MethodHandle methodHandle)
+    {
+        _methodHandle = Objects.requireNonNull(methodHandle);
+        _numParams = methodHandle.type().parameterCount();
+        _parameters = new Object[_numParams];
+        _boundParams = new Boolean[_numParams];
+    }
+
+    private int getInternalIndex(int idx)
+    {
+        int index = 0;
+        for (int i = 0; i < _numParams; i++)
+        {
+            if (_boundParams[i] != Boolean.TRUE)
+            {
+                if (index == idx)
+                    return i;
+                index++;
+            }
+        }
+
+        throw new IndexOutOfBoundsException(idx);
+    }
+
+    @Override
+    public Object invoke(Object... args) throws Throwable
+    {
+        insertArguments(false, 0, args);
+        Object o = _methodHandle.invokeWithArguments(_parameters);
+        if (_returnFilter != null)
+            o = _returnFilter.invoke(o);
+        return (_returnType == null) ? o : _returnType.cast(o);
+    }
+
+    @Override
+    public JettyMethodHandle bindTo(Object arg, int idx)
```

Comment:
Is this style of binding to a specific index really necessary?   MethodHandle doesn't support that and you can only bind from the first args.

If we supported the same semantic, then you'd only need a single index to track the binding (rather than a boolean array) and you could use efficient arraycopy operations to move args into the array on every invoke rather than a loop with conditional tests over two arrays!